File: AccessPath Source Code/iOS/Access Path/Custom Objects/GoogleMapsArcGISAdapter.swift
//
//  GoogleMapsArcGISAdapter.swift
//  Access Path
//
//  Created by Alan Barker on 12/11/19.
//  Copyright © 2019 pathVu. All rights reserved.
//

import Foundation
import GoogleMaps
import SwiftyJSON

protocol GoogleMapsArcGISAdapter {
    var coordinates: [CLLocationCoordinate2D] { get set }
    func setupAndRender()
    func queryAndRender()
}

class TransitLayer : MultiMarkerLayer {
    override var type: String { MapLayers.transitType }
    override var outputFields: String { "objectid,stop_name,stop_type" }
    init(map: GMSMapView) {
        super.init(url: MapLayers.transitLayerURL!, map: map)
    }
}

class CurbrampLayer : SingleMarkerLayer {
    override var type: String { MapLayers.curbrampType }
    override var outputFields: String { "detectable_warning,globalid,objectid" }
    init(map: GMSMapView) {
        super.init(url: MapLayers.curbRampsLayerURL!, map: map)
    }
}

class SidewalkLayer : PathLayer {
    override var type: String { MapLayers.sidewalkType }
    override var outputFields: String { "objectid,segment_rai,street_name" }
    init(map: GMSMapView) {
        super.init(url: MapLayers.sidewalkLayerURL!, map: map)
    }
}

class CrowdsourceLayer : SingleMarkerLayer {
    override var type: String { MapLayers.crowdsourceType }
    override var outputFields: String { "cid,cname" }
    init(map: GMSMapView) {
        super.init(url: MapLayers.crowdsourcingLayerURL!, map: map)
    }
}

class BaseLayer : GoogleMapsArcGISAdapter {
    var coordinates = [CLLocationCoordinate2D]()
    var url: URL
    var map: GMSMapView
    var type: String { "" }
    var outputFields: String { "" }
    
    init(url: URL, map: GMSMapView) {
        self.url = url
        self.map = map
    }
    
    func getQueryItems(topRight: CLLocationCoordinate2D, bottomLeft: CLLocationCoordinate2D) -> [URLQueryItem] {
        return [
            URLQueryItem(name: "outSR", value: "4326"),
            URLQueryItem(name: "returnGeometry", value: "true"),
            URLQueryItem(name: "inSR", value: "4326"),
            URLQueryItem(name: "returnDistinctValues", value: "false"),
            URLQueryItem(name: "maxAllowableOffset", value: "0.000000"),
            URLQueryItem(name: "spatialRel", value: "esriSpatialRelEnvelopeIntersects"),
            URLQueryItem(name: "geometryType", value: "esriGeometryEnvelope"),
            URLQueryItem(name: "outFields", value: outputFields),
            URLQueryItem(name: "geometry", value: "{xmin:\(topRight.longitude),ymin:\(bottomLeft.latitude),xmax:\(bottomLeft.longitude),ymax=\(topRight.latitude)}"),
            URLQueryItem(name: "f", value: "json"),
            URLQueryItem(name: "returnZ", value: "false"),
            URLQueryItem(name: "returnM", value: "false")
        ]
    }

    func setupAndRender() {
    }
    
    func queryAndRender() {
    }
}
class MultiMarkerLayer : BaseLayer {
    var markerIcons = [String:UIImage]()
    override func setupAndRender() {
        var query = URLComponents(string: url.absoluteString)!
        query.queryItems = [URLQueryItem(name: "f", value: "json")]
        URLSession.shared.dataTask(with: query.url!, completionHandler: { data, response, error in
            if error != nil {
                return
            }
            let json = JSON(data!)
            for item in json["drawingInfo"]["renderer"]["uniqueValueInfos"].arrayValue {
                let imageString = item["symbol"]["imageData"].stringValue
                if let itemData = Data(base64Encoded: imageString, options:[]){
                    let icon = UIImage(data: itemData, scale: 0.8)
                    self.markerIcons[item["value"].stringValue] = icon
                }
            }
            self.queryAndRender()
        }).resume()
    }
    
    override func queryAndRender() {
        DispatchQueue.main.async {
            let topRight = self.map.projection.visibleRegion().farRight
            let bottomLeft = self.map.projection.visibleRegion().nearLeft
            
            var query = URLComponents(string: self.url.absoluteString)!
            query.path += "/query"
            query.queryItems = self.getQueryItems(topRight: topRight, bottomLeft: bottomLeft)
            URLSession.shared.dataTask(with: query.url!, completionHandler: { data, response, error in
                DispatchQueue.main.async {
                    if error != nil {
                        return
                    }
                    let json = JSON(data!)
                    for item in json["features"].arrayValue {
                        let coord = CLLocationCoordinate2D(latitude:item["geometry"]["y"].doubleValue, longitude: item["geometry"]["x"].doubleValue)
                        let marker = GMSMarker(position: coord)
                        marker.icon = self.markerIcons[item["attributes"]["stop_type"].stringValue]
                        var userData = item["attributes"]
                        userData["type"].stringValue = self.type
                        marker.userData = userData
                        marker.map = self.map
                    }
                }
            }).resume()
        }
    }
}

class SingleMarkerLayer : BaseLayer {
    var icon: UIImage?
    override func setupAndRender() {
        var query = URLComponents(string: url.absoluteString)!
        query.queryItems = [URLQueryItem(name: "f", value: "json")]
        URLSession.shared.dataTask(with: query.url!, completionHandler: { data, response, error in
            if error != nil {
                return
            }
            let json = JSON(data!)
            let imageString = json["drawingInfo"]["renderer"]["symbol"]["imageData"].stringValue
            if let imageData = Data(base64Encoded: imageString, options:[]) {
                self.icon = UIImage(data: imageData, scale: 0.8)
            }
            self.queryAndRender()
        }).resume()
    }
    
    override func queryAndRender() {
        if let icon = self.icon {
            DispatchQueue.main.async {
                let topRight = self.map.projection.visibleRegion().farRight
                let bottomLeft = self.map.projection.visibleRegion().nearLeft
                
                var query = URLComponents(string: self.url.absoluteString)!
                query.path += "/query"
                query.queryItems = self.getQueryItems(topRight: topRight, bottomLeft: bottomLeft)
                
                URLSession.shared.dataTask(with: query.url!, completionHandler: { data, response, error in
                    DispatchQueue.main.async {
                        if error != nil {
                            return
                        }
                        let json = JSON(data!)
                        for item in json["features"].arrayValue {
                            let coord = CLLocationCoordinate2D(latitude: item["geometry"]["y"].doubleValue, longitude: item["geometry"]["x"].doubleValue)
                            // only add coordinates to property for hazard layer
                            if self.type == MapLayers.crowdsourceType {
                                self.coordinates.append(coord)
                            }
                            let marker = GMSMarker(position: coord)
                            var userData = item["attributes"]
                            userData["type"].stringValue = self.type
                            marker.userData = userData
                            marker.icon = icon
                            marker.map = self.map
                        }
                    }
                }).resume()
            }
        }
    }
}

class PathLayer : BaseLayer {
    var colorValues = [String:Float]()
    var lines = [Any]()
    override func setupAndRender() {
        var query = URLComponents(string: url.absoluteString)!
        query.queryItems = [URLQueryItem(name: "f", value: "json")]
        URLSession.shared.dataTask(with: query.url!, completionHandler: { data, response, error in
            if error != nil {
                return
            }
            let json = JSON(data!)

            for info in json["drawingInfo"]["renderer"]["classBreakInfos"].arrayValue {
                self.colorValues[info["label"].stringValue] = info["classMaxValue"].floatValue
            }
            self.queryAndRender()
        }).resume()
    }
    
    override func queryAndRender() {
        DispatchQueue.main.async {
            let topRight = self.map.projection.visibleRegion().farRight
            let bottomLeft = self.map.projection.visibleRegion().nearLeft
            
            var query = URLComponents(string: self.url.absoluteString)!
            query.path += "/query"
            query.queryItems = self.getQueryItems(topRight: topRight, bottomLeft: bottomLeft)
            
            URLSession.shared.dataTask(with: query.url!, completionHandler: { data, response, error in
                DispatchQueue.main.async {
                    if error != nil {
                        return
                    }
                    self.lines.removeAll()
                    let json = JSON(data!)
                    for item in json["features"].arrayValue {
                        for path in item["geometry"]["paths"].arrayValue {
                            let p = GMSMutablePath()
                            for points in path.arrayValue {
                                let pArr = points.arrayValue
                                p.add(CLLocationCoordinate2D(latitude: pArr[1].doubleValue, longitude:pArr[0].doubleValue))
                            }
                            let polyline = GMSPolyline(path: p)
                            
                            guard let greenVal = self.colorValues["0.000000 - 1.000000"],
                                let yellowVal = self.colorValues["1.000001 - 3.000000"],
                                let redVal = self.colorValues["3.000001 - 224.500000"]
                                //let blackVal = self.colorValues["Unrated"]
                            else {
                                return
                            }
                            
                            if item["attributes"]["segment_rai"].floatValue == 0.0 {
                                //continue
                                polyline.strokeColor = UIColor.black
                            }
                            else if item["attributes"]["segment_rai"].floatValue < greenVal {
                                polyline.strokeColor = UIColor.green
                            }
                            else if item["attributes"]["segment_rai"].floatValue < yellowVal {
                                polyline.strokeColor = UIColor.yellow
                            }
                            else if item["attributes"]["segment_rai"].floatValue < redVal {
                                polyline.strokeColor = UIColor.red
                            }
                            /*else if item["attributes"].label == "Unrated" {
                                polyline.strokeColor = UIColor.black
                            }*/
                            var userData = item["attributes"]
                            userData["type"].stringValue = self.type
                            polyline.userData = userData
                            polyline.strokeWidth = 3
                            polyline.isTappable = true
                            polyline.map = self.map
                            
                        }
                    }
                }
            }).resume()
        }
    }
}

File: AccessPath Source Code/Android/app/src/main/java/com/pathvu/accesspath2020/Util/GoogleMapsArcGISAdapter.kt
package com.pathvu.accesspath2020.Util

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Build
import android.util.Base64
import android.util.Log
import androidx.annotation.RequiresApi
import com.android.volley.Cache
import com.android.volley.DefaultRetryPolicy
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.*
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.pathvu.accesspath2020.model.CurbRamp
import com.pathvu.accesspath2020.model.Transit
import org.json.JSONObject
import java.net.URLEncoder

/**
 * Get transit, curb ramp and sidewalk information from ArcGIS and display these three layer on the map
 */
abstract class GoogleMapsArcGISAdapter(val url: StringBuilder, val map: GoogleMap, val context: Context) {
    abstract var layerType: String
    abstract var outputField: String

    abstract fun setupAndRender()
    abstract fun queryAndRender()

    fun getQueryItems(topRight: LatLng, bottomLeft: LatLng) {
        val bound = "{xmin:" + topRight.longitude.toString() + ",ymin:" + bottomLeft.latitude.toString() + ",xmax:" + bottomLeft.longitude.toString() + ",ymax=" + topRight.latitude.toString() + "}"
        url.append("outSR=" + URLEncoder.encode("4326", "utf8"))
        url.append("&returnGeometry=" + URLEncoder.encode("true", "utf8"))
        url.append("&inSR=" + URLEncoder.encode("4326", "utf8"))
        url.append("&returnDistinctValues=" + URLEncoder.encode("false", "utf8"))
        url.append("&maxAllowableOffset=" + URLEncoder.encode("0.000000", "utf8"))
        url.append("&spatialRel=" + URLEncoder.encode("esriSpatialRelEnvelopeIntersects", "utf8"))
        url.append("&geometryType=" + URLEncoder.encode("esriGeometryEnvelope", "utf8"))
        url.append("&outFields=" + URLEncoder.encode(outputField, "utf8"))
        url.append("&geometry=" + URLEncoder.encode(bound, "utf8"))
        url.append("&f=" + URLEncoder.encode("json", "utf8"))
        url.append("&returnZ=" + URLEncoder.encode("false", "utf8"))
        url.append("&returnM=" + URLEncoder.encode("false", "utf8"))
    }
}



open class MultiMarkerLayer(url: StringBuilder, map: GoogleMap, context: Context, var queue: RequestQueue, var locationMap: HashMap<Marker, String>) : GoogleMapsArcGISAdapter(url, map, context) {
    override lateinit var layerType: String
    override lateinit var outputField: String
    var originalURL: String = url.toString()
    var markerIcon = HashMap<String, BitmapDescriptor>()

    override fun setupAndRender() {
        val transitURL = "$originalURL?f=json"
        val stringRequest = @RequiresApi(Build.VERSION_CODES.O)
        object : StringRequest(
            Method.GET, transitURL.toString(),
            Response.Listener<String> { response ->
                val responseString = response.toString()
                val resJSON = JSONObject(responseString).getJSONObject("drawingInfo").getJSONObject("renderer").getJSONArray("uniqueValueInfos")
                for(i in 0 until resJSON.length()) {
                    val imageString = resJSON.getJSONObject(i).getJSONObject("symbol").getString("imageData")
                    val imageData = Base64.decode(imageString, Base64.DEFAULT)
                    val icon = BitmapFactory.decodeByteArray(imageData, 0, imageData.size)
                    val iconDescriptor = BitmapDescriptorFactory.fromBitmap(icon)
                    markerIcon[resJSON.getJSONObject(i).getString("value")] = iconDescriptor
                    if(!icon.isRecycled) {
                        icon.recycle()
                        System.gc()
                    }
                }
                queryAndRender()
            },
            Response.ErrorListener { error ->
                println("error: $error") })
        {

        }
        stringRequest.retryPolicy = DefaultRetryPolicy(50000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        queue.add(stringRequest)
    }


    override fun queryAndRender() {
        val topRight = map.projection.visibleRegion.farRight
        val bottomLeft = map.projection.visibleRegion.nearLeft

        url.append("/query?")
        getQueryItems(topRight, bottomLeft)

        val stringRequest = @RequiresApi(Build.VERSION_CODES.O)
        object : StringRequest(
            Method.GET, url.toString(),
            Response.Listener<String> { response ->
                val responseString = response.toString()
                val resJSON = JSONObject(responseString).getJSONArray("features")
                for(i in 0 until resJSON.length()) {
                    val curJson = resJSON.getJSONObject(i)
                    val coord = LatLng(curJson.getJSONObject("geometry").getString("y").toDouble(), curJson.getJSONObject("geometry").getString("x").toDouble())
                    val attr = curJson.getJSONObject("attributes")
                    val stopType = attr.getString("stop_type")
                    val stopName = attr.getString("stop_name")
                    val direction = attr.getString("direction")
                    val shelter = attr.getString("shelter")
                    val routes = attr.getString("routes")
                    val transit = Transit(stopName, stopType, coord, direction, routes, shelter)
                    val markerOptions = MarkerOptions().position(coord)
                    markerOptions.icon(markerIcon[stopType])
                    val marker = map.addMarker(markerOptions)
                    marker.tag =transit
                    locationMap[marker] = MapLayers.transitType
                }
            },
            Response.ErrorListener { error ->
                println("error: $error") })
        {

        }
        stringRequest.retryPolicy = DefaultRetryPolicy(50000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        queue.add(stringRequest)
    }
}



open class SingleMarkerLayer(url: StringBuilder, map: GoogleMap, context: Context, var queue: RequestQueue, var locationMap: java.util.HashMap<Marker, String>) : GoogleMapsArcGISAdapter(url, map, context) {
    override lateinit var layerType: String
    override lateinit var outputField: String
    var originalURL: String = url.toString()
    lateinit var markerIcon: BitmapDescriptor

    override fun setupAndRender() {
        val transitURL = "$originalURL?f=json"
        val stringRequest = @RequiresApi(Build.VERSION_CODES.O)
        object : StringRequest(
            Method.GET, transitURL,
            Response.Listener<String> { response ->
                val responseString = response.toString()
                val imageString = JSONObject(responseString).getJSONObject("drawingInfo").getJSONObject("renderer").getJSONObject("symbol").getString("imageData")
                val imageData = Base64.decode(imageString, Base64.DEFAULT)
                val icon = BitmapFactory.decodeByteArray(imageData, 0, imageData.size)
                markerIcon = BitmapDescriptorFactory.fromBitmap(icon)
                queryAndRender()
                if(!icon.isRecycled) {
                    icon.recycle()
                    System.gc()
                }
            },
            Response.ErrorListener { error ->
                println("error: $error") })
        {

        }
        stringRequest.retryPolicy = DefaultRetryPolicy(50000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        queue.add(stringRequest)
    }


    override fun queryAndRender() {
        val topRight = map.projection.visibleRegion.farRight
        val bottomLeft = map.projection.visibleRegion.nearLeft

        url.append("/query?")
        getQueryItems(topRight, bottomLeft)

        val stringRequest = @RequiresApi(Build.VERSION_CODES.O)
        object : StringRequest(
            Method.GET, url.toString(),
            Response.Listener<String> { response ->
                val responseString = response.toString()
//                println("single part " + toPrettyFormat(responseString))
                val resJSON = JSONObject(responseString).getJSONArray("features")
                for(i in 0 until resJSON.length()) {
                    val curJson = resJSON.getJSONObject(i)
                    val coord = LatLng(curJson.getJSONObject("geometry").getString("y").toDouble(), curJson.getJSONObject("geometry").getString("x").toDouble())
                    val markerOptions = MarkerOptions().position(coord)
                    val attr = curJson.getJSONObject("attributes")
                    val lippage = attr.getString("lippage")
                    val userSlope = attr.getString("user_slope")
                    val overallCondition = attr.getString("overall_condition")
                    val imaeUrl = attr.getString("imageurl")
                    val curbRamp = CurbRamp(coord, lippage, userSlope, overallCondition, imaeUrl)
                    markerOptions.icon(markerIcon)
                    val marker = map.addMarker(markerOptions)
                    marker.tag = curbRamp
                    locationMap[marker] = MapLayers.curbRampType
                }
            },
            Response.ErrorListener { error ->
                println("error: $error") })
        {

        }
        stringRequest.retryPolicy = DefaultRetryPolicy(50000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        queue.add(stringRequest)
    }
}



open class PathMarkerLayer(url: StringBuilder, map: GoogleMap, context: Context, var queue: RequestQueue) : GoogleMapsArcGISAdapter(url, map, context) {
    override lateinit var layerType: String
    override lateinit var outputField: String
    var originalURL: String = url.toString()
    var colorVal = HashMap<String, Float>()

    override fun setupAndRender() {
        val transitURL = "$originalURL?f=json"
        val stringRequest = @RequiresApi(Build.VERSION_CODES.O)
        object : StringRequest(
            Method.GET, transitURL,
            Response.Listener<String> { response ->
                val responseString = response.toString()
                val resJson = JSONObject(responseString).getJSONObject("drawingInfo").getJSONObject("renderer").getJSONArray("classBreakInfos")
                for(i in 0 until resJson.length()) {
                    val curJson = resJson.getJSONObject(i)
                    colorVal[curJson.getString("label")] = curJson.getString("classMaxValue").toFloat()
                }
                queryAndRender()
            },
            Response.ErrorListener { error ->
                println("error: $error") })
        {

        }
        stringRequest.retryPolicy = DefaultRetryPolicy(50000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        queue.add(stringRequest)
    }


    override fun queryAndRender() {
        val topRight = map.projection.visibleRegion.farRight
        val bottomLeft = map.projection.visibleRegion.nearLeft

        url.append("/query?")
        getQueryItems(topRight, bottomLeft)

        val stringRequest = @RequiresApi(Build.VERSION_CODES.O)
        object : StringRequest(
            Method.GET, url.toString(),
            Response.Listener<String> { response ->
                val responseString = response.toString()
                val resJSON = JSONObject(responseString).getJSONArray("features")
                for(i in 0 until resJSON.length()) {
                    val curJson = resJSON.getJSONObject(i)
                    val paths = curJson.getJSONObject("geometry").getJSONArray("paths")
                    for(j in 0 until paths.length()) {
                        val curPath = paths.getJSONArray(j)
                        val p = PolylineOptions()
                        for(k in 0 until curPath.length()) {
                            val path = curPath[k].toString()
                            val curPathInfo = path.substring(1, path.length - 1).split(',')
                            p.add(LatLng(curPathInfo[1].toDouble(), curPathInfo[0].toDouble()))
                        }
                        val greenVal = colorVal["0.000000 - 1.000000"]
                        val yellowVal = colorVal["1.000001 - 3.000000"]
                        val redVal = colorVal["3.000001 - 224.500000"]
                        if(!curJson.getJSONObject("attributes").isNull("segment_rai")) {
                            if (curJson.getJSONObject("attributes").getString("segment_rai").toFloat() == 0.0.toFloat()) {
//                                p.color(Color.parseColor("#000000"))
//                                p.width(3.0f)
//                                map.addPolyline(p)
                            }
                            else if (curJson.getJSONObject("attributes").getString("segment_rai").toFloat() < greenVal!!) {
                                p.color(Color.parseColor("#00ff00"))
                                p.width(3.0f)
                                map.addPolyline(p)
                            }
                            else if (curJson.getJSONObject("attributes").getString("segment_rai").toFloat() < yellowVal!!) {
                                p.color(Color.parseColor("#ffff00"))
                                p.width(3.0f)
                                map.addPolyline(p)
                            }
                            else if (curJson.getJSONObject("attributes").getString("segment_rai").toFloat() < redVal!!) {
                                p.color(Color.parseColor("#ff0000"))
                                p.width(3.0f)
                                map.addPolyline(p)
                            }
                        } else {
                            p.color(Color.parseColor("#000000"))
                            p.width(3.0f)
                            map.addPolyline(p)
                        }
                    }
                }
            },
            Response.ErrorListener { error ->
                println("error: $error") })
        {

        }
        stringRequest.retryPolicy = DefaultRetryPolicy(50000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        queue.add(stringRequest)
    }
}


class TransitLayer(url: StringBuilder, map: GoogleMap, context: Context, queue: RequestQueue, locationMap: java.util.HashMap<Marker, String>) : MultiMarkerLayer(url, map, context, queue, locationMap) {
    override var layerType: String = MapLayers.transitType
    override var outputField: String = "objectid,stop_name,stop_type,direction,shelter,routes"
}

class CurbRampLayerLayer(url: StringBuilder, map: GoogleMap, context: Context, queue: RequestQueue, locationMap: java.util.HashMap<Marker, String>) : SingleMarkerLayer(url, map, context, queue, locationMap) {
    override var layerType: String = MapLayers.curbRampType
    override var outputField: String = "detectable_warning,globalid,objectid,lippage,user_slope,overall_condition,imageurl"
}

class SidewalkLayer(url: StringBuilder, map: GoogleMap, context: Context, queue: RequestQueue) : PathMarkerLayer(url, map, context, queue) {
    override var layerType: String = MapLayers.sidewalkType
    override var outputField: String = "segment_rai,objectid,street_name"
}
